### ISSUE TITLE
Fails to compile free-5.2

### ISSUE THREAD

**Issue by hasufell** [tweag/ghc-wasm-miso-examples#10]
**Describe the bug**

```
Failed to build free-5.2.
Build log (
/home/hasufell/.cabal/logs/ghc-9.10.0.20240313/free-5.2-91deb61bd419de4c7ce7d3b73925dab74b44de19f780c97f91a818cb2c9f80f1.log
):
Configuring library for free-5.2..
Preprocessing library for free-5.2..
Building library for free-5.2..
[ 1 of 19] Compiling Control.Alternative.Free ( src/Control/Alternative/Free.hs, dist/build/Control/Alternative/Free.o )
[ 2 of 19] Compiling Control.Alternative.Free.Final ( src/Control/Alternative/Free/Final.hs, dist/build/Control/Alternative/Free/Final.o )
[ 3 of 19] Compiling Control.Applicative.Free ( src/Control/Applicative/Free.hs, dist/build/Control/Applicative/Free.o )
[ 4 of 19] Compiling Control.Applicative.Free.Fast ( src/Control/Applicative/Free/Fast.hs, dist/build/Control/Applicative/Free/Fast.o )
[ 5 of 19] Compiling Control.Applicative.Free.Final ( src/Control/Applicative/Free/Final.hs, dist/build/Control/Applicative/Free/Final.o )
[ 6 of 19] Compiling Control.Applicative.Trans.Free ( src/Control/Applicative/Trans/Free.hs, dist/build/Control/Applicative/Trans/Free.o )
[ 7 of 19] Compiling Control.Comonad.Cofree.Class ( src/Control/Comonad/Cofree/Class.hs, dist/build/Control/Comonad/Cofree/Class.o )
[ 8 of 19] Compiling Control.Comonad.Cofree ( src/Control/Comonad/Cofree.hs, dist/build/Control/Comonad/Cofree.o )
[ 9 of 19] Compiling Control.Comonad.Trans.Cofree ( src/Control/Comonad/Trans/Cofree.hs, dist/build/Control/Comonad/Trans/Cofree.o )
[10 of 19] Compiling Control.Comonad.Trans.Coiter ( src/Control/Comonad/Trans/Coiter.hs, dist/build/Control/Comonad/Trans/Coiter.o )
[11 of 19] Compiling Control.Monad.Free.Class ( src/Control/Monad/Free/Class.hs, dist/build/Control/Monad/Free/Class.o )
[12 of 19] Compiling Control.Monad.Free.TH ( src/Control/Monad/Free/TH.hs, dist/build/Control/Monad/Free/TH.o )

src/Control/Monad/Free/TH.hs:110:44: error: [GHC-83865]
    • Couldn't match type ‘Pat’ with ‘ArgPat’
      Expected: Name -> ArgPat
        Actual: Name -> Pat
    • In the first argument of ‘map’, namely ‘VarP’
      In the first argument of ‘LamE’, namely ‘(map VarP xs)’
      In the second argument of ‘Captured’, namely
        ‘(LamE (map VarP xs) (nonUnaryTupE $ map VarE xs))’
    |
110 |           return $ Captured tup (LamE (map VarP xs) (nonUnaryTupE $ map VarE xs))
    |                                            ^^^^

src/Control/Monad/Free/TH.hs:221:30: error: [GHC-83865]
    • Couldn't match type ‘Pat’ with ‘ArgPat’
      Expected: [ArgPat]
        Actual: [Pat]
    • In the first argument of ‘Clause’, namely ‘pat’
      In the expression: Clause pat (NormalB $ AppE (VarE liftF) fval) []
      In the second argument of ‘FunD’, namely
        ‘[Clause pat (NormalB $ AppE (VarE liftF) fval) []]’
    |
221 |     , [ FunD opName [ Clause pat (NormalB $ AppE (VarE liftF) fval) [] ] ] ]
    |                              ^^^
[13 of 19] Compiling Control.Monad.Trans.Free ( src/Control/Monad/Trans/Free.hs, dist/build/Control/Monad/Trans/Free.o )
[14 of 19] Compiling Control.Monad.Free ( src/Control/Monad/Free.hs, dist/build/Control/Monad/Free.o )
[15 of 19] Compiling Control.Monad.Free.Church ( src/Control/Monad/Free/Church.hs, dist/build/Control/Monad/Free/Church.o )
[16 of 19] Compiling Control.Monad.Trans.Free.Ap ( src/Control/Monad/Trans/Free/Ap.hs, dist/build/Control/Monad/Trans/Free/Ap.o )
[17 of 19] Compiling Control.Monad.Free.Ap ( src/Control/Monad/Free/Ap.hs, dist/build/Control/Monad/Free/Ap.o )
[18 of 19] Compiling Control.Monad.Trans.Free.Church ( src/Control/Monad/Trans/Free/Church.hs, dist/build/Control/Monad/Trans/Free/Church.o )
[19 of 19] Compiling Control.Monad.Trans.Iter ( src/Control/Monad/Trans/Iter.hs, dist/build/Control/Monad/Trans/Iter.o )
Error: cabal: Failed to build free-5.2 (which is required by
```

**To Reproduce**

* set up `ghc-wasm-meta` first:
  * `git clone https://gitlab.haskell.org/ghc/ghc-wasm-meta.git`
  * `cd ghc-wasm-meta`
  * `export SKIP_GHC=yes`
  * `./setup.sh`
  * `source ~/.ghc-wasm/env`
* `ghcup install ghc --set wasm32-wasi-9.10.0.20240313 -- --host=x86_64-linux --with-intree-gmp --with-system-libffi`
* `cabal build --with-ghc=wasm32-wasi-ghc --with-ghc-pkg=wasm32-wasi-ghc-pkg --with-hsc2hs=wasm32-wasi-hsc2hs`

**Expected behavior**

Compiles.

**Environment**
 - OS name + version: Fedora 39
 - Version of the code: 3045defa7f5a06e8a8c4f8a54ea3f84169c1fac3

**Additional context**

*Not* using nix (and not interested to try).

**Comment by hasufell** [comment]
It seems to compile with wasm32-wasi-9.10.0.20240412, which I just added to GHCup: https://github.com/haskell/ghcup-metadata/commit/d86cdc6ad361215fe2198825cfb36e058c57733f

**Comment by TerrorJack** [comment]
Closed via #12.
